app/src/main/java/com/amaze/filemanager/asynchronous/asynctasks/LoadFilesListTask.java

```diff
@@ -142,56 +143,91 @@ public LoadFilesListTask(
-          list = mainFragment.addToSmb(smbFile, path, showHiddenFiles);
-          openmode = OpenMode.SMB;
-        } catch (SmbAuthException e) {
-          if (!e.getMessage().toLowerCase().contains("denied")) {
+        ArrayList<LayoutElementParcelable> smbCache = mainFragmentViewModel.getFromListCache(path);
+        openmode = OpenMode.SMB;
+        if (smbCache != null) {
+          list = smbCache;
+        } else {
+          try {
+            SmbFile[] smbFile = hFile.getSmbFile(5000).listFiles();
+            list = mainFragment.addToSmb(smbFile, path, showHiddenFiles);
+          } catch (SmbAuthException e) {
+            if (!e.getMessage().toLowerCase().contains("denied")) {
+              mainFragment.reauthenticateSmb();
+            }
+            LOG.warn("failed to load smb list, authentication issue", e);
+            return null;
+          } catch (SmbException | NullPointerException e) {
+            LOG.warn("Failed to load smb files for path: " + path, e);
             mainFragment.reauthenticateSmb();
+            return null;
           }
-          LOG.warn("failed to load smb list, authentication issue", e);
-          return null;
-        } catch (SmbException | NullPointerException e) {
-          LOG.warn("Failed to load smb files for path: " + path, e);
-          mainFragment.reauthenticateSmb();
-          return null;
+          mainFragmentViewModel.putInCache(path, list);
         }
         break;
       case SFTP:
         HybridFile sftpHFile = new HybridFile(OpenMode.SFTP, path);
-
-        list = new ArrayList();
-
-        sftpHFile.forEachChildrenFile(
-            context,
-            false,
-            file -> {
-              if (!(dataUtils.isFileHidden(file.getPath())
-                  || file.isHidden() && !showHiddenFiles)) {
-                LayoutElementParcelable elem = createListParcelables(file);
-                if (elem != null) {
-                  list.add(elem);
+        ArrayList<LayoutElementParcelable> sftpCache = mainFragmentViewModel.getFromListCache(path);
+        if (sftpCache != null) {
+          list = sftpCache;
+        } else {
+          list = new ArrayList();
+          sftpHFile.forEachChildrenFile(
+              context,
+              false,
+              file -> {
+                if (!(dataUtils.isFileHidden(file.getPath())
+                    || file.isHidden() && !showHiddenFiles)) {
+                  LayoutElementParcelable elem = createListParcelables(file);
+                  if (elem != null) {
+                    list.add(elem);
+                  }
                 }
-              }
-            });
+              });
+          mainFragmentViewModel.putInCache(path, list);
+        }
         break;
       case CUSTOM:
         switch (Integer.parseInt(path)) {
           case 0:
-            list = listImages();
+            if (mainFragmentViewModel.getStorageImages() == null) {
```

Comment:
Code seems very repeated, the switch probably should be generalized?

Reply:
Fixed

Reply:
Fixed

Reply:
Fixed